File: FiveCardStud/cpp/poker.cpp
#include<fstream>
#include<iostream>
#include<cstring>
#include<string>
#include<cstdlib>
#include<ctime>
using namespace std;

const int HIGH_CARD = 0;
const int PAIR = 1;
const int TWO_PAIR = 2;
const int THREE_OF_A_KIND = 3;
const int STRAIGHT = 4;
const int FLUSH = 5;
const int FULL_HOUSE = 6;
const int FOUR_OF_A_KIND = 7;
const int STRAIGHT_FLUSH = 8;
const int ROYAL_FLUSH = 9;
const int VALID = -1;//used for Card.value for duplicateCard

const string faces[] = {"2","3","4","5","6","7","8","9","10","J","Q","K","A"};
const string handRanks[] = {"High Card","Pair","Two Pair","Three of a Kind","Straight","Flush","Full House","Four of a Kind","Straight Flush","Royal Flush"};

class Card{
public:
	int value;
	char suit;
	Card( int value, char suit ) : value(value), suit(suit){}
	Card(){}

	int suitRank(){
		switch( suit ){
			case 'D':
				return 3;
			case 'C':
				return 2;
			case 'H':
				return 1;
			case 'S':
				return 0;
		}
		return -1;
	}
	static int getValue( char c ){
		if( c >= '2' && c <= '9' )
			return ( c - '0' ) - 2;
		switch( c ){
			case 'J': return 9;
			case 'Q': return 10;
			case 'K': return 11;
			case 'A': return 12;
		}
		return -1;
	}
	string toString(){
		return faces[value]+suit;
	}
};

class Hand{
	public:
	Card myCards[5];
	bool markedCards[4][13];
	int numEachCard[13];
	int handRank = -1;
	int flushSuit = -1;//suit of the flush
	int kickerSuit = -1;//suit of the kicker card in a two pair
	int kickerValue = -1;
	int pairHighSuit = -1;//suit of the highest value card not in the pair
	int valueOfTriple = -1;//value of the card appearing 3 times in a full house or three of a kind
	int fourKindVal = -1;
	int pairValue = -1;
	int twoPairValue = -1;
	int doubleCardVal = -1;//for full house

	Hand(){}

	void aceLowStraightCheck(){//if an ace low straight is present, the hand is shifted to put the ace in the last spot and 5 at the first
		if( myCards[0].value != 12 )
			return;
		if( myCards[1].value != 3 )
			return;
		if( myCards[2].value != 2 )
			return;
		if( myCards[3].value != 1 )
			return;
		if( myCards[4].value != 0 )
			return;
		Card holdAce = myCards[0];
		for( int x = 0; x < 3; x++ )
			myCards[x] = myCards[x+1];
		myCards[4] = holdAce;
	}
	void getNumEachCard(){
		for( int x = 0; x < 52; x++ )
			numEachCard[x] = 0;
		for( int x =0; x < 5; x++ )
			numEachCard[myCards[x].value]++;
	}

	static void sortHands( Hand hands[], int indexTracker[] ){
		for( int x = 0; x < 6; x++ )
			indexTracker[x] = x;
		bool anotherPassNeeded = true;
		while( anotherPassNeeded ){
			anotherPassNeeded = false;
			for( int x =0; x < 5; x++ )
				if( hands[x].losesTo( hands[x+1] ) ){
					anotherPassNeeded = true;
					Hand tempHand = hands[x];
					hands[x] = hands[x+1];
					hands[x+1] = tempHand;

					int tempIndex = indexTracker[x];
					indexTracker[x] = indexTracker[x+1];
					indexTracker[x+1] = tempIndex;
				}
		}
	}

	bool losesTo( Hand hand2 ){
		if( handRank != hand2.handRank )
			return handRank < hand2.handRank;
		if( handRank == HIGH_CARD ){
			for( int x = 0; x < 5; x++ )
				if( myCards[x].value != hand2.myCards[x].value )
					return myCards[x].value < hand2.myCards[x].value;
			return myCards[0].suitRank() < hand2.myCards[0].suitRank();
		}
		if( handRank == PAIR ){
			if( pairValue != hand2.pairValue )
				return pairValue < hand2.pairValue;
			int firstIndex = 0;
			int secondIndex = 0;
			while( firstIndex != 5 && secondIndex != 5 ){
				while( myCards[firstIndex].value == pairValue )
					firstIndex++;
				while( hand2.myCards[secondIndex].value == hand2.pairValue )
                                               secondIndex++;
				if( myCards[firstIndex].value != hand2.myCards[secondIndex].value )
					return myCards[firstIndex].value < hand2.myCards[secondIndex].value;
				firstIndex++;
				secondIndex++;
			}
			return kickerSuit < hand2.kickerSuit;
		}
		if( handRank == TWO_PAIR ){
			if( twoPairValue != hand2.twoPairValue )
				return twoPairValue < hand2.twoPairValue;
			if( pairValue != hand2.pairValue )
				return pairValue < hand2.pairValue;
			if( kickerValue != hand2.kickerValue )
				return kickerValue < hand2.kickerValue;
			return kickerSuit < hand2.kickerSuit;
		}
		if( handRank == THREE_OF_A_KIND ){
			if( valueOfTriple != hand2.valueOfTriple )
				return valueOfTriple < hand2.valueOfTriple;
			for( int x = 0 ; x < 5; x++ )
				if( myCards[x].value != hand2.myCards[x].value )
					return myCards[x].value < hand2.myCards[x].value;
			return myCards[0].suitRank() < hand2.myCards[0].suitRank();
		}
		if(handRank == STRAIGHT ){
			if( myCards[0].value != hand2.myCards[0].value )
				return myCards[0].value < hand2.myCards[0].value;
			return myCards[0].suitRank() < hand2.myCards[0].suitRank();
		}
		if( handRank == FLUSH ){
			for( int x = 0; x < 5; x++ )
				if( myCards[x].value != hand2.myCards[x].value )
					return myCards[x].value < hand2.myCards[x].value;
			return flushSuit < hand2.flushSuit;
		}
		if( handRank == FULL_HOUSE ){
			if( valueOfTriple != hand2.valueOfTriple )
				return valueOfTriple < hand2.valueOfTriple;
			if( doubleCardVal != hand2.doubleCardVal )
				return doubleCardVal < hand2.doubleCardVal;
			for( int x = 0; x < 5; x++ )
				if( myCards[x].suitRank() != hand2.myCards[x].suitRank() )
					return myCards[x].suitRank() < hand2.myCards[x].suitRank();
		}
		if( handRank == FOUR_OF_A_KIND ){
			if( fourKindVal != hand2.fourKindVal )
				return fourKindVal < hand2.fourKindVal;
			for( int x = 0; x < 5; x++ )
				if( myCards[x].value != hand2.myCards[x].value )
					return myCards[x].value < hand2.myCards[x].value;
			for( int x = 0; x < 5; x++ )//the if statement here is guaranteed to trigger once
				if( myCards[x].suitRank() != hand2.myCards[x].suitRank() )
					return myCards[x].suitRank() < hand2.myCards[x].suitRank();
		}
		if( handRank == STRAIGHT_FLUSH ){
			if( myCards[0].value != hand2.myCards[0].value )
				return myCards[0].value < hand2.myCards[0].value;
			return myCards[0].suitRank() < hand2.myCards[0].suitRank();
		}
		if( handRank == ROYAL_FLUSH )
			return flushSuit < hand2.flushSuit;
		return true; //unreachable but neccessary for compiler
	}

	void gradeHand(){
		if( royalFlush() ){
			handRank = ROYAL_FLUSH;
			return;
		}else if( straightFlush() ){
			handRank = STRAIGHT_FLUSH;
			return;
		}else if( fourOfAKind() ){
			handRank = FOUR_OF_A_KIND;
			return;
		}else if( fullHouse() ){
			handRank = FULL_HOUSE;
			return;
		}else if( flush() ){
			handRank = FLUSH;
			return;
		}else if( straight() ){
			handRank = STRAIGHT;
			return;
		}else if( threeOfAKind() ){
			handRank = THREE_OF_A_KIND;
			return;
		}else if( twoPair() ){
			handRank = TWO_PAIR;
			return;
		}else if( pair() ){
			handRank = PAIR;
			return;
		}
		handRank = HIGH_CARD;
	}

	bool royalFlush(){
		for( int x = 0; x < 4; x++ )
			for(int y = 8; y < 14; y++){//indexes in markedCards which can represent a card in the flush
				if( y == 13 ){//reachable only if a royal flush appears
					flushSuit = x;
					return true;
				}
				if( ! markedCards[x][y] )
					break;
			}
		return false;
	}

	bool straightFlush(){
		for( int x = 0; x < 4; x++ ){//does not handle ace low straight flushes
			int curConsecutive = 0;
			for( int y = 0; y < 13; y++ ){
				if( curConsecutive == 5 ){
					flushSuit = x;
					return true;
				}
				if( markedCards[x][y] ){
					curConsecutive++;
				}else
					curConsecutive = 0;
			}
		}
		if( myCards[4].value != 12 )//fifth card is an ace
			return false;
		char potentialSuit = myCards[0].suit;
		if( myCards[3].value != 0 || myCards[3].suit != potentialSuit )//fourth card is a 2
			return false;
		if( myCards[2].value != 1 || myCards[2].suit != potentialSuit )//third card is a 3
			return false;
		if( myCards[1].value != 2 || myCards[1].suit != potentialSuit )//second card is a 4
			return false;
		if( myCards[0].value != 3 || myCards[0].suit != potentialSuit )//first card is a 5
			return false;
		flushSuit = myCards[0].suitRank();
		return true;
	}

	bool fourOfAKind(){
		for( int x = 0; x < 13; x++ )
			if( numEachCard[x] == 4 ){
				fourKindVal = x;
				return true;
			}
		return false;
	}

	bool fullHouse(){
		bool tripleFound = false;
		bool doubleFound = false;
		for( int x = 0; x < 13; x++ ){
			if( numEachCard[x] == 3 ){
				valueOfTriple = x;
				tripleFound = true;
			}else if( numEachCard[x] == 2 ){
				doubleFound = true;
				doubleCardVal = x;
			}
			if( tripleFound && doubleFound )
				return true;
		}
		return false;
	}

	bool flush(){
		for( int x = 0; x < 4; x++ ){
			int suitCount = 0;
			for( int y = 0; y < 13; y++ ){
				if( suitCount == 5 ){
					flushSuit = x;
					return true;
				}
				if( markedCards[x][y] )
					suitCount++;
			}
		}
		return false;
	}

	bool straight(){
                int curConsecutive = 0;
                for( int y = 0; y < 13; y++ ){//
                        if( curConsecutive == 5 )
                                return true;
                        if( numEachCard[y] >=1 ){
                               	curConsecutive++;
			}else
                        	curConsecutive = 0;
         	}
		if( myCards[4].value != 12 )//fifth card is an ace
                        return false;
                if( myCards[3].value != 0 )//fourth card is a 2
                        return false;
                if( myCards[2].value != 1 )//third card is a 3
                        return false;
                if( myCards[1].value != 2 )//second card is a 4
                        return false;
                if( myCards[0].value != 3 )//first card is a 5
                        return false;
                return true;
	}

	bool threeOfAKind(){
                for( int x = 0 ; x < 13; x++ )
			if( numEachCard[x] == 3 ){
				valueOfTriple = x;
				return true;
			}
		return false;

	}

	bool twoPair(){
		bool pairFound = false;
		for( int x = 0; x < 13; x++ )
			if( numEachCard[x] > 1 ){
				if( pairFound ){
					twoPairValue = x;
					markKicker();
					if( twoPairValue < pairValue ){//makes sure the higher value is in twoPairValue
						int temp = pairValue;
						pairValue = twoPairValue;
						twoPairValue = temp;
					}
					return true;
				}else{
					pairValue = x;
					pairFound = true;
				}
			}
		return false;
	}

	void markKicker(){
		for( int x = 0; x < 13; x++ )
			if( numEachCard[x] == 1 )//only triggered when the card is the odd card out of two pairs
				for( int y = 0; y < 5; y++ )
					if( myCards[y].value == x ){
						kickerSuit = myCards[y].suitRank();
						kickerValue = myCards[y].value;
						return;
					}
	}

	bool pair(){
		for( int x = 0; x < 13; x++ ){
			if( numEachCard[x] > 1 ){//true if pair is found
				int suitHighCardNoPair = -1;
				int valueHighCardNoPair = -1;
				for( int y = 0; y < 5; y++ )//finds max value card not part of the pair
					if( (numEachCard[ myCards[y].value ] == 1) && valueHighCardNoPair < myCards[y].value ){
						valueHighCardNoPair = myCards[y].value;
						suitHighCardNoPair = myCards[y].suitRank();
					}
				kickerSuit = suitHighCardNoPair;
				kickerValue = valueHighCardNoPair;
				pairValue = x;
				return true;
			}
		}
		return false;
	}

	void markCards(){
		for( int x =0; x < 4; x++ )
			for( int y = 0; y < 13; y++ )
				markedCards[x][y] = false;
                for( int x = 0; x < 5; x++ ){
                        Card card = myCards[x];
                        markedCards[card.suitRank()][card.value] = true;
                }
	}

};

class poker{
public:
	static string getFullFile( string fileName ){
		ifstream scan(fileName);
		string fullFile = "";
		string line = "";
		while( getline( scan, line ) )
			fullFile += line+'\n';
		fullFile = fullFile.substr(0, 119);
		scan.close();
		return fullFile;
	}

	static void printRawHands( Hand hands[] ){
		cout << "\n\n";
		for( int x = 0; x < 6; x++ ){
			Hand curHand = hands[x];
			cout << "Hand " << (x+1) << ": ";
			for( int y = 0; y < 5; y++ ){
				Card curCard = curHand.myCards[y];
				cout << "(" << curCard.value << ", " << curCard.suit << "), ";
			}
			cout << endl;
		}
		cout << "\n\n";
	}

	static void getFileHands( string fileName, Hand hands[] ){
		ifstream scan(fileName);
		string line;
		int index = 0;
		while( getline( scan, line ) ){
		//for( int x = 0; x < 6; x++ ){
		//	getline( scan, line );
			string rawCards[5];
			Hand hand;
			for( int y= 0; y < 5; y++ )//fill rawCards with strings
				rawCards[y] = line.substr(y*4,3);
			for( int y = 0; y < 5; y++ ){
				string cardString = rawCards[y];
				Card card(-1,'z');
				if( cardString[0] == '1' ){
					card.value = 8;
 					card.suit = cardString[2];
				}else{
					card.value = Card::getValue( cardString[1] );
					card.suit = cardString[2];
				}
				hand.myCards[y] = card;
			}
			hands[index++] = hand;
		}
		scan.close();
	}

	static void printDeckAndHands( Hand hands[], Card cards[], bool random, bool duplicateCard, string fileName, string fullFile ){
		cout << "*** P O K E R  H A N D  A N A L Y Z E R ***\n" << endl;
		if( random ){
			cout << "*** USING RANDOMIZED DECK OF CARDS ***\n\n*** Shuffled 52 card deck:" << endl;
			cout << cards[0].toString() << " ";
			for( int x = 1; x < 52; x++ ){
                               	if( !random && x % 5 == 0 )
                                       	cout << " ";
                               	cout << cards[x].toString() << " ";
				if( x % 13 == 0 )
                                       	cout << endl;
			}
		}else
			cout << "*** USING TEST DECK ***\n\n*** File: " << fileName << endl << fullFile;
		if( duplicateCard )
			return;
		cout << "\n\n*** Here are the six hands..." << endl;
		for( int a = 0; a < 6; a++){
			for( int y = 0; y < 5; y++ )
				cout << hands[a].myCards[y].toString() << " ";
			cout << endl;
		}
		if( !random )
			return;
		cout << "\n*** Here is what remains in the deck..." << endl;
		for( int x = 30; x < 52; x++ )
			cout << cards[x].toString() << " ";
		cout << endl;
	}
	static void printResults( Hand hands[], Hand originalHands[], int indexTracker[] ){
		cout << "\n --- WINNING HAND ORDER ---" << endl;
		for( int x = 0 ; x < 6 ; x++ ){
			int origIndexToPrint = indexTracker[x];
			cout << " " << originalHands[origIndexToPrint].myCards[0].toString();
			for( int y = 1; y < 5; y++ ){
				string cardString = originalHands[origIndexToPrint].myCards[y].toString();
				if( originalHands[origIndexToPrint].myCards[0].value == 8 && y == 1 && cardString.length() == 2 )
					cardString = " " + cardString;
				else if( cardString.length() == 2 )
					cardString = "  " + cardString;
				else
					cardString = " " + cardString;
				cout << cardString;
			}
		cout << " - " << handRanks[hands[x].handRank] << endl;
		}
	}

	static void copyHands( Hand hands[], Hand originalHands[] ){
		for( int x = 0; x < 6; x++ ){
			//Hand newHand();
			//originalHands[x] = newHand;
			for( int y = 0; y < 5; y++ ){
				Card newCard( hands[x].myCards[y].value, hands[x].myCards[y].suit);
				originalHands[x].myCards[y] = newCard;
			}
		}
	}
	static void getRandCards( Card cards[] ){
		char suits[] = {'H','D','S','C'};
		int counter = 0;
		for( int x = 0; x < 4; x++ )
			for( int y = 0; y < 13; y++ ){
				Card newCard( y, suits[x]);
				cards[ counter++ ] = newCard;
			}
		shuffle( cards );
	}

	static void shuffle( Card cards[] ){
		srand( time(0) );//seeds the RNG
		for( int y = 0; y < 3; y++ ){//runs a lot of times because the randomness is not great
			for( int x = 51; x > 0; x-- ){
				int index = rand() % (x+1);
				Card temp = cards[index];
				cards[index] = cards[x];
				cards[x] = temp;
			}
		}
	}

	static void generateRandHands( Card cards[], Hand hands[] ){
		int curHand = 0;
		int iteration = 0;
		for( int x = 0; x < 6; x++ ){
			Hand newHand;
			hands[x] = newHand;
		}
		for( int x = 0 ; x < 30; x++ ){
			hands[curHand].myCards[iteration] = cards[x];
			curHand++;
			if( curHand == 6 ){
				curHand = 0;
				iteration++;
			}
		}
	}

	static void sortEach( Hand hands[] ){
		for( int a = 0; a < 6; a++ ){
			bool anotherPassNeeded = true;
			while( anotherPassNeeded ){
				anotherPassNeeded = false;
				for( int x = 0; x < 4; x++ ){
					if( hands[a].myCards[x].value < hands[a].myCards[x+1].value ){
						Card temp = hands[a].myCards[x];
						hands[a].myCards[x] = hands[a].myCards[x+1];
						hands[a].myCards[x+1] = temp;
						anotherPassNeeded = true;
					}
				}
			}
			hands[a].aceLowStraightCheck();
		}
	}

	static Card checkDuplicateCards( Hand hands[], Card cards[] ){
		bool cardsFound[4][13];
		for( int x = 0 ; x < 4; x++ )
			for( int y = 0; y < 13; y++ )
				cardsFound[x][y] = false;
		for( int x = 0; x < 6; x++ )
			for( int y = 0; y < 5; y++ ){
				int suitRank = hands[x].myCards[y].suitRank();
				int value = hands[x].myCards[y].value;
				if( cardsFound[ suitRank][ value ] )
					return hands[x].myCards[y];
				else
					cardsFound[ suitRank][ value ] = true;
			}
		Card validCard(VALID, 'z');
		return validCard;
	}

};

int main( int argc, char* argv[] ){
	cout << endl;
	Card cards[52];
	Hand hands[6];
	string fileName = "";
	string fullFile = "";
	bool random = argc == 1;
	if( random ){
		poker::getRandCards(cards);
		poker::generateRandHands( cards, hands );
	}else{
		string temp = argv[1];
		fileName = "../handsets/"+temp;//cur directory is the langauge directory
		fullFile = poker::getFullFile(fileName);
		poker::getFileHands( fileName, hands );//this method sets cards to have every card in the hand
	}
	Card duplicateCard( VALID, 'z' );
	if( argc == 2 )
		duplicateCard = poker::checkDuplicateCards( hands, cards );
	poker::printDeckAndHands(hands, cards, random, duplicateCard.value != VALID, fileName, fullFile);
	if( duplicateCard.value != VALID ){
		cout << "\n\n*** ERROR - DUPLICATED CARD FOUND IN DECK ***" << endl;
		cout << "\n\n*** DUPLICATE: " << duplicateCard.toString() << " ***" << endl;
		exit(0);
	}
	Hand originalHands[6];
	poker::copyHands( hands, originalHands );
	poker::sortEach( hands );
	for( int x =0; x < 6; x++ ){
		hands[x].markCards();
		Hand fixSegFault;
		if( x != 5 )
			fixSegFault = hands[x+1];//for some reason this call to NumEachCard sometimes likes to change the values of all the cards in this hand to 0 and all the suitranks to -1
		hands[x].getNumEachCard();
		if( x != 5 )
			hands[x+1] = fixSegFault;
		hands[x].gradeHand();
	}
	int indexTracker[6];
	Hand::sortHands( hands, indexTracker );
	poker::printResults( hands, originalHands, indexTracker );
	return 0;
}

File: FiveCardStud/rust/src/poker.rs
//mod card;
use crate::hand::Hand;
use crate::card::Card;
use std::env;
use std::fs;
use rand::prelude::*;
use rand::Rng;
use rand::thread_rng;

const HIGH_CARD : u8 = 0;
const PAIR : u8 = 1;
const TWO_PAIR : u8 = 2;
const THREE_OF_A_KIND : u8 = 3;
const STRAIGHT : u8 = 4;
const FLUSH : u8 = 5;
const FULL_HOUSE : u8 = 6;
const FOUR_OF_A_KIND : u8 = 7;
const STRAIGHT_FLUSH : u8 = 8;
const ROYAL_FLUSH : u8 = 9;
const HAND_RANKS: [&str; 10] = ["High Card","Pair","Two Pair","Three of a Kind", "Straight", "Flush", "Full House", "Four Of A Kind", "Straight Flush", "Royal Flush"];

pub fn printResults( hands : [Hand; 6], originalHands : [Hand; 6], indexTracker : [usize;6] ){
    println!("\n --- WINNING HAND ORDER ---");
    for x in 0usize..6usize{
        let mut origIndexToPrint = indexTracker[x];
        print!(" {}", originalHands[origIndexToPrint].myCards[0].to_string());
        for y in 1usize..5usize{
            let mut handHelp = originalHands[origIndexToPrint];
            let mut cardString = String::from("");
            if y == 2 && handHelp.myCards[1].value == 8 && handHelp.myCards[0].value == 8 {
                cardString.push_str(" ");
            } else if handHelp.myCards[0].value == 8 && y == 1 && originalHands[origIndexToPrint].myCards[y].value != 8{
                cardString.push_str(" ");
            } else if originalHands[origIndexToPrint].myCards[y].value != 8{
                cardString.push_str("  ");
            } else {
                cardString.push_str(" ");
            }
            cardString.push_str(handHelp.myCards[y].to_string().as_str());
            print!("{}",cardString);
        }
        println!(" - {}", HAND_RANKS[hands[x].handRank as usize]);
    }
}

pub fn sortHands( hands : &mut [Hand;6] ) -> [usize; 6]{
    let mut indexTracker : [usize; 6] = [0;6];
    for x in 0usize..6usize{
        indexTracker[x] = x;
    }
    let mut anotherPassneeded = true;
    while anotherPassneeded{
        anotherPassneeded = false;
        for x in 0usize..5usize{
            if hands[x].losesTo(hands[x+1]){
                anotherPassneeded = true;
                let tempHand = hands[x];
                hands[x] = hands[x+1];
                hands[x+1] = tempHand;

                let tempIndex = indexTracker[x];
                indexTracker[x] = indexTracker[x+1];
                indexTracker[x+1] = tempIndex;
            }
        }
    }
    return indexTracker;
}

pub fn gradeHand( hand : &mut Hand ){
    if hand.royalFlush(){
        hand.handRank = ROYAL_FLUSH;
    } else if hand.straightFlush(){
        hand.handRank = STRAIGHT_FLUSH;
    } else if hand.fourOfAKind(){
        hand.handRank = FOUR_OF_A_KIND;
    } else if hand.fullHouse(){
        hand.handRank = FULL_HOUSE;
    } else if hand.flush(){
        hand.handRank = FLUSH;
    } else if hand.straight(){
        hand.handRank = STRAIGHT;
    } else if hand.threeOfAKind(){
        hand.handRank = THREE_OF_A_KIND;
    } else if hand.twoPair(){
        hand.handRank = TWO_PAIR;
    } else if hand.pair(){
        hand.handRank = PAIR;
    } else{
        hand.handRank = HIGH_CARD;
    }
}

pub fn checkDupCards( hands : [Hand;6] ) -> Card{
    let mut cardsFound : [[bool;13];4] = [[false;13];4];
    for i in 0..6{
        for j in 0..5{
            let mut card = hands[i].myCards[j];
            let mut sRank = card.suit_rank();
            let mut value = card.value;
            if cardsFound[sRank as usize][value as usize]{
                return card;
            }else{
                cardsFound[sRank as usize][value as usize] = true;
            }
        }
    }
    return Card::new(-1,'z');
}

pub fn padCardArray( smallCards : [Card; 30] ) -> [Card; 52]{  
    let mut cards : [Card;52] = [Card::default(); 52];
    for x in 0..30{
        cards[x as usize] = smallCards[x as usize];
    }
    cards
}

pub fn getRandCards() -> [Card; 52] {
    let mut cards: [Card; 52] = [Card::default(); 52];
    let suits = ['H','D','S','C'];
    for x in 0..4{
        for y in 0..13{
            let card = Card::new(y, suits[x as usize]);
            cards[(x*13+y) as usize] = card;
        }
    }

    let mut rng = thread_rng();

    for i in (0..cards.len()).rev() {
      let j = rng.gen_range(0..=i);
      cards.swap(i, j);
    }
    cards
}

pub fn getRandHands(cards : [Card; 52]) -> [Hand; 6] {
let mut hands: [Hand; 6] = [Hand::default(); 6];
let mut curHand = 0;
let mut numCard = 0;
/*for card in cards{
    println!("{}", card.to_string() );
}*/
for x in 0..30{
    hands[curHand as usize ].myCards[numCard as usize] = cards[x as usize];
    curHand += 1;
    if curHand == 6 {
        curHand = 0;
        numCard += 1;
    }
}
hands
}

pub fn markCards( hand : &mut Hand ){
    for x in 0..5{
        let mut sRank = hand.myCards[x as usize].suit_rank();
        hand.markedCards[sRank as usize][hand.myCards[x as usize].value as usize] = true;
    }
}

pub fn getNumEachCard( hand : &mut Hand ){
    for x in 0..5{
        hand.numEachCard[ hand.myCards[x as usize].value as usize] += 1;
    }
}

pub fn sortEach( hands : &mut [Hand;6]){
    for x in 0..6{
        let mut anotherPassneeded = true;
        while anotherPassneeded{
            anotherPassneeded = false;
            for y in 0..4{
                if hands[x as usize].myCards[y as usize].value < hands[x as usize].myCards[(y+1) as usize].value{
                    anotherPassneeded = true;
                    let mut temp = hands[x as usize].myCards[y as usize];
                    hands[x as usize].myCards[y as usize] = hands[x as usize].myCards[(y+1) as usize];
                    hands[x as usize].myCards[(y+1) as usize] = temp;
                }
            }
        }
        aceLowStraightCheck( &mut hands[x as usize]);
    }
}

pub fn aceLowStraightCheck( hand : &mut Hand ){
    if hand.myCards[0].value != 12{
        return;
    } else if hand.myCards[1].value != 3{
        return;
    } else if hand.myCards[2].value != 2{
        return;
    } else if hand.myCards[3].value != 1{
        return;
    } else if hand.myCards[4].value != 0{
        return
    }
    let mut temp = hand.myCards[0];
    for x in 0..4{
        hand.myCards[x as usize] = hand.myCards[(x+1) as usize];
    }
    hand.myCards[4] = temp;

}

pub fn printDeckAndHands( hands : [Hand;6], cards : [Card;52], random : bool, dupPresent : bool, fullFile : String){
    println!("{}","\n*** P O K E R  H A N D  A N A L Y Z E R ***\n");
    if random{
        println!("{}","\n*** USING RANDOMIZED DECK OF CARDS ***\n\n*** Shuffled 52 card deck:");
        print!("{} ", cards[0].to_string() );
        for x in 1..52{
            print!("{} ", cards[x as usize].to_string());
            if x % 13 == 0{
                println!("");
            }
        }
        println!("");
    } else{
        let args: Vec<String> = env::args().collect();
        let name = &args[1];
        print!("{}{}\n{}","\n*** USING TEST DECK ***\n\n*** File: ../../handsets/", name, fullFile);
    }
    if dupPresent {
        return;
    }
    println!("{}","\n*** Here are the six hands...");
    for x in 0..6{
        for y in 0..5{
            print!("{} ",hands[x as usize].myCards[y as usize].to_string());
        }
        println!("");
    }
    if !random{
        return;
    }
    println!("{}","\n*** Here is what remains in the deck...");
    for x in 30..52{
        print!("{} ",cards[x as usize].to_string());
    }
    println!("");
}

pub fn getFullFile() -> String{
    let args: Vec<String> = env::args().collect();
    let name = &args[1];
    let fullPath = "../../handsets/".to_string() + name;
    let filename = String::from(fullPath);

    let contents = fs::read_to_string(filename)
        .expect("Something went wrong reading the file");
    contents
}

pub fn getFileCards( fullFile : String ) -> [Card; 30]{
    let mut cards = [Card::default(); 30];
    let lines: Vec<&str> = fullFile.split("\n").collect();
    let mut counter = 0;
    for line in &lines{
        if line.is_empty(){
            continue;
        }
        let mut lineStrings: Vec<String> = line.split(",")
                                .map(|s| s.to_string())
                                .collect();
        for x in 0..5{
            lineStrings[x as usize] = lineStrings[x as usize].trim().to_string();
            //print!(":{}:", lineStrings[counter as usize]);
            //counter += 1;
            if lineStrings[x as usize].len() == 2{
                cards[counter] = Card::new(Card::value_from_char(
                    lineStrings[x as usize].chars().nth(0).unwrap()),  lineStrings[x as usize].chars().nth(1).unwrap())
                    
            }else{
                cards[counter] = Card::new(8, lineStrings[x as usize].chars().nth(2).unwrap());
            }
            counter += 1;
        }
        //println!("");
    }
    cards
}

pub fn getFileHands( cards : [Card;52] ) -> [Hand; 6] {
    let mut hands = [Hand::default(); 6];
    for i in 0..6{
        for j in 0..5{
            hands[i].myCards[j] = cards[(i*5+j) as usize];
        }
    }
    hands
}